File: test-rigs/go-test-rig/usecase/usecase.go
package usecase

import (
	"github.com/jkjamies/omnivore/go-test-rig/model"
	"github.com/jkjamies/omnivore/go-test-rig/repository"
	"github.com/jkjamies/omnivore/go-test-rig/validation"
)

// AddTask creates and stores a new task with validation.
func AddTask(repo repository.TaskRepository, id uint64, title string, priority model.Priority, description string, tags []string) (*model.Task, error) {
	if err := validation.ValidateTitle(title); err != nil {
		return nil, err
	}
	for _, tag := range tags {
		if err := validation.ValidateTag(tag); err != nil {
			return nil, err
		}
	}

	task := model.NewTask(id, title, priority)
	task.Description = validation.SanitizeDescription(description)
	task.Tags = tags

	if err := repo.Add(task); err != nil {
		return nil, err
	}
	return task, nil
}

// ToggleTask flips a task's completion status.
func ToggleTask(repo repository.TaskRepository, id uint64) (*model.Task, error) {
	task, err := repo.GetByID(id)
	if err != nil {
		return nil, err
	}
	task.ToggleCompleted()
	if err := repo.Update(task); err != nil {
		return nil, err
	}
	return task, nil
}

// RemoveTask deletes a task by ID.
func RemoveTask(repo repository.TaskRepository, id uint64) (*model.Task, error) {
	return repo.Remove(id)
}

// GetTasks returns all tasks, optionally filtered by search query.
func GetTasks(repo repository.TaskRepository, query string) []*model.Task {
	all := repo.GetAll()
	if query == "" {
		return all
	}
	var result []*model.Task
	for _, t := range all {
		if t.MatchesSearch(query) {
			result = append(result, t)
		}
	}
	return result
}

// GetTasksByPriority groups tasks by priority (Critical first, Low last).
func GetTasksByPriority(repo repository.TaskRepository) []PriorityGroup {
	all := repo.GetAll()
	priorities := []model.Priority{
		model.PriorityCritical,
		model.PriorityHigh,
		model.PriorityMedium,
		model.PriorityLow,
	}
	var groups []PriorityGroup
	for _, p := range priorities {
		var tasks []*model.Task
		for _, t := range all {
			if t.Priority == p {
				tasks = append(tasks, t)
			}
		}
		if len(tasks) > 0 {
			groups = append(groups, PriorityGroup{Priority: p, Tasks: tasks})
		}
	}
	return groups
}

// PriorityGroup holds tasks grouped by priority.
type PriorityGroup struct {
	Priority model.Priority
	Tasks    []*model.Task
}

// TaskStats holds summary statistics.
type TaskStats struct {
	Total          int
	Completed      int
	Pending        int
	Actionable     int
	CompletionRate float64
}

// GetStats computes summary statistics.
func GetStats(repo repository.TaskRepository) TaskStats {
	all := repo.GetAll()
	total := len(all)
	completed := 0
	actionable := 0
	for _, t := range all {
		if t.Completed {
			completed++
		}
		if t.IsActionable() {
			actionable++
		}
	}
	rate := 0.0
	if total > 0 {
		rate = float64(completed) / float64(total)
	}
	return TaskStats{
		Total:          total,
		Completed:      completed,
		Pending:        total - completed,
		Actionable:     actionable,
		CompletionRate: rate,
	}
}

File: test-rigs/rust-test-rig/core/src/usecase.rs
use crate::model::{Priority, Task, TaskResult};
use crate::repository::TaskRepository;
use crate::validation;

/// Add a new task with validation.
pub fn add_task(
    repo: &mut dyn TaskRepository,
    id: u64,
    title: &str,
    priority: Priority,
    description: Option<&str>,
    tags: Vec<String>,
) -> TaskResult<Task> {
    validation::validate_title(title)?;
    for tag in &tags {
        validation::validate_tag(tag)?;
    }

    let desc = description
        .map(validation::sanitize_description)
        .unwrap_or_default();

    let task = Task::new(id, title, priority)
        .with_description(desc)
        .with_tags(tags);

    repo.add(task.clone())?;
    Ok(task)
}

/// Toggle a task's completion status.
pub fn toggle_task(repo: &mut dyn TaskRepository, id: u64) -> TaskResult<Task> {
    let mut task = repo.get_by_id(id)?;
    task.toggle_completed();
    repo.update(task.clone())?;
    Ok(task)
}

/// Remove a task by ID.
pub fn remove_task(repo: &mut dyn TaskRepository, id: u64) -> TaskResult<Task> {
    repo.remove(id)
}

/// Get all tasks, optionally filtered by a search query.
pub fn get_tasks(repo: &dyn TaskRepository, query: Option<&str>) -> Vec<Task> {
    let all = repo.get_all();
    match query {
        Some(q) if !q.is_empty() => all.into_iter().filter(|t| t.matches_search(q)).collect(),
        _ => all,
    }
}

/// Get tasks grouped by priority.
pub fn get_tasks_by_priority(repo: &dyn TaskRepository) -> Vec<(Priority, Vec<Task>)> {
    let all = repo.get_all();
    let priorities = [
        Priority::Critical,
        Priority::High,
        Priority::Medium,
        Priority::Low,
    ];

    priorities
        .into_iter()
        .map(|p| {
            let tasks: Vec<Task> = all.iter().filter(|t| t.priority == p).cloned().collect();
            (p, tasks)
        })
        .filter(|(_, tasks)| !tasks.is_empty())
        .collect()
}

/// Get summary statistics.
pub fn get_stats(repo: &dyn TaskRepository) -> TaskStats {
    let all = repo.get_all();
    let total = all.len();
    let completed = all.iter().filter(|t| t.completed).count();
    let actionable = all.iter().filter(|t| t.is_actionable()).count();
    TaskStats {
        total,
        completed,
        pending: total - completed,
        actionable,
        completion_rate: if total > 0 {
            completed as f64 / total as f64
        } else {
            0.0
        },
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct TaskStats {
    pub total: usize,
    pub completed: usize,
    pub pending: usize,
    pub actionable: usize,
    pub completion_rate: f64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::repository::InMemoryTaskRepository;

    #[test]
    fn add_task_success() {
        let mut repo = InMemoryTaskRepository::new();
        let task = add_task(&mut repo, 1, "New task", Priority::Medium, None, vec![]).unwrap();
        assert_eq!(task.title, "New task");
        assert_eq!(repo.count(), 1);
    }

    #[test]
    fn add_task_with_description_and_tags() {
        let mut repo = InMemoryTaskRepository::new();
        let task = add_task(
            &mut repo,
            1,
            "Tagged task",
            Priority::High,
            Some("A  detailed   description"),
            vec!["backend".into(), "urgent".into()],
        )
        .unwrap();
        assert_eq!(task.description, "A detailed description");
        assert_eq!(task.tags.len(), 2);
    }

    #[test]
    fn add_task_invalid_title() {
        let mut repo = InMemoryTaskRepository::new();
        assert!(add_task(&mut repo, 1, "", Priority::Low, None, vec![]).is_err());
    }

    #[test]
    fn add_task_invalid_tag() {
        let mut repo = InMemoryTaskRepository::new();
        assert!(add_task(
            &mut repo,
            1,
            "Task",
            Priority::Low,
            None,
            vec!["bad tag".into()]
        )
        .is_err());
    }

    #[test]
    fn get_tasks_all() {
        let mut repo = InMemoryTaskRepository::new();
        add_task(&mut repo, 1, "Alpha", Priority::Low, None, vec![]).unwrap();
        add_task(&mut repo, 2, "Beta", Priority::High, None, vec![]).unwrap();
        assert_eq!(get_tasks(&repo, None).len(), 2);
    }

    #[test]
    fn get_tasks_filtered() {
        let mut repo = InMemoryTaskRepository::new();
        add_task(&mut repo, 1, "Fix login", Priority::High, None, vec![]).unwrap();
        add_task(&mut repo, 2, "Add signup", Priority::Medium, None, vec![]).unwrap();
        let results = get_tasks(&repo, Some("login"));
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].title, "Fix login");
    }

    #[test]
    fn get_stats_empty() {
        let repo = InMemoryTaskRepository::new();
        let stats = get_stats(&repo);
        assert_eq!(stats.total, 0);
        assert_eq!(stats.completion_rate, 0.0);
    }

    #[test]
    fn get_stats_mixed() {
        let mut repo = InMemoryTaskRepository::new();
        add_task(&mut repo, 1, "Done", Priority::Low, None, vec![]).unwrap();
        toggle_task(&mut repo, 1).unwrap();
        add_task(&mut repo, 2, "Pending", Priority::High, None, vec![]).unwrap();
        let stats = get_stats(&repo);
        assert_eq!(stats.total, 2);
        assert_eq!(stats.completed, 1);
        assert_eq!(stats.pending, 1);
        assert_eq!(stats.actionable, 1);
        assert_eq!(stats.completion_rate, 0.5);
    }

    // Intentionally NOT testing: toggle_task, remove_task, get_tasks_by_priority
}
